File: backend/src/schemas/channel.py
"""Channel schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ChannelCreate(BaseModel):
    """Schema for creating a channel."""

    name: str
    description: str | None = None
    workspace_id: int


class ChannelUpdate(BaseModel):
    """Schema for updating a channel."""

    name: str | None = None
    description: str | None = None


class ChannelResponse(BaseModel):
    """Schema for channel response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    workspace_id: int
    owner_id: int
    created_at: datetime


class ChannelMemberResponse(BaseModel):
    """Schema for channel member response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    channel_id: int
    user_id: int
    role: str
    joined_at: datetime


class ChannelWithMembers(BaseModel):
    """Schema for channel with members."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    workspace_id: int
    owner_id: int
    created_at: datetime
    members: list[ChannelMemberResponse] = []


class ChannelListResponse(BaseModel):
    """Schema for channel list response."""

    items: list[ChannelResponse]
    total: int

File: frontend/src/api/channel.ts
import axios from './axios'

export interface Channel {
  id: number
  name: string
  description: string | null
  workspace_id: number
  owner_id: number
  created_at: string
}

export interface ChannelMember {
  id: number
  channel_id: number
  user_id: number
  role: string
  joined_at: string
}

export interface CreateChannelData {
  name: string
  description?: string
  workspace_id: number
}

export interface UpdateChannelData {
  name?: string
  description?: string
}

export interface ChannelListResponse {
  items: Channel[]
  total: number
}

export const channelApi = {
  getChannels: async (workspaceId?: number): Promise<ChannelListResponse> => {
    const params = workspaceId ? { workspace_id: workspaceId } : {}
    const response = await axios.get<ChannelListResponse>('/api/channels', { params })
    return response.data
  },

  getChannel: async (channelId: number): Promise<Channel> => {
    const response = await axios.get<Channel>(`/api/channels/${channelId}`)
    return response.data
  },

  createChannel: async (data: CreateChannelData): Promise<Channel> => {
    const response = await axios.post<Channel>('/api/channels', data)
    return response.data
  },

  updateChannel: async (channelId: number, data: UpdateChannelData): Promise<Channel> => {
    const response = await axios.put<Channel>(`/api/channels/${channelId}`, data)
    return response.data
  },

  deleteChannel: async (channelId: number): Promise<void> => {
    await axios.delete(`/api/channels/${channelId}`)
  },

  joinChannel: async (channelId: number): Promise<Channel> => {
    const response = await axios.post<Channel>(`/api/channels/${channelId}/join`)
    return response.data
  },

  leaveChannel: async (channelId: number): Promise<void> => {
    await axios.post(`/api/channels/${channelId}/leave`)
  },

  getMembers: async (channelId: number): Promise<ChannelMember[]> => {
    const response = await axios.get<ChannelMember[]>(`/api/channels/${channelId}/members`)
    return response.data
  },

  removeMember: async (channelId: number, userId: number): Promise<void> => {
    await axios.delete(`/api/channels/${channelId}/members/${userId}`)
  },
}
